pkg/ruler/ruler.go

```diff
@@ -550,3 +563,71 @@ func (r *Ruler) Rules(ctx context.Context, in *RulesRequest) (*RulesResponse, er
 
 	return &RulesResponse{Groups: groupDescs}, nil
 }
+
+// updateRules will force each ruler to update the rules for the userID.
+// This is called right after a user updates his config, this way they don't need to
+// wait until the refresh interval to see their config changes.
+func (r *Ruler) updateRules(userID string) {
+	ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
+	defer cancel()
+
+	if !r.cfg.EnableSharding {
+		r.loadRules(ctx, userID)
+		return
+	}
+
+	rulers, err := r.ring.GetAll(ring.Read)
+	if err != nil {
+		level.Error(r.logger).Log("msg", "unable to read ring for updateRules", "err", err)
+		return
+	}
+
+	ctx = user.InjectOrgID(ctx, userID)
+	ctx, err = user.InjectIntoGRPCRequest(ctx)
+	if err != nil {
+		level.Error(r.logger).Log("msg", "unable to inject user ID into grpc request", "err", err)
+		return
+	}
+
+	var wg sync.WaitGroup
+
+	for _, rlr := range rulers.Ingesters {
+		wg.Add(1)
+
+		go func(logger log.Logger, addr string, wg *sync.WaitGroup) {
+			defer wg.Done()
+
+			dialOpts, err := r.cfg.ClientTLSConfig.GetGRPCDialOptions()
+			if err != nil {
+				level.Error(r.logger).Log("msg", "unable to get dail options", "err", err)
+				return
+			}
+			conn, err := grpc.Dial(addr, dialOpts...)
+			if err != nil {
+				level.Error(r.logger).Log("msg", "unable to dial gRPC connection", "err", err)
+				return
+			}
+			cc := NewRulerClient(conn)
+			_, err = cc.ReloadRules(ctx, nil)
+
+			if err != nil {
+				level.Error(r.logger).Log("msg", "err reloading rules", "err", err, "ruler", addr)
+				return
+			}
+
+		}(r.logger, rlr.Addr, &wg)
+	}
+
+	wg.Wait()
+}
+
+// ReloadRules implements the rules service.
+func (r *Ruler) ReloadRules(ctx context.Context, _ *ReloadRulesRequest) (*emptypb.Empty, error) {
+	userID, err := user.ExtractOrgID(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("no user id found in context")
+	}
+
+	r.loadRules(ctx, userID)
```

Comment:
```suggestion
	go r.loadRules(ctx, userID)
```

Might as well run this in the background since no error is returned at this point.